plugin/trino-base-jdbc/src/main/java/io/trino/plugin/jdbc/ReusingConnectionFactory.java

```diff
@@ -0,0 +1,205 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.plugin.jdbc;
+
+import com.google.common.cache.CacheBuilder;
+import io.trino.collect.cache.NonLoadableCache;
+import io.trino.collect.cache.SafeCaches;
+import io.trino.spi.TrinoException;
+import io.trino.spi.connector.ConnectorSession;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.annotation.concurrent.ThreadSafe;
+import javax.inject.Inject;
+
+import java.sql.Connection;
+import java.sql.SQLException;
+import java.time.Duration;
+import java.util.Map;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.cache.RemovalCause.EXPLICIT;
+import static io.trino.plugin.jdbc.JdbcErrorCode.JDBC_ERROR;
+import static java.util.Objects.requireNonNull;
+
+@ThreadSafe
+public final class ReusingConnectionFactory
+        implements ConnectionFactory, QueryEventListener
+{
+    @GuardedBy("this")
+    private final Map<String, Connection> connections;
+    private final ConnectionFactory delegate;
+
+    @Inject
+    public ReusingConnectionFactory(@ForLazyConnectionFactory ConnectionFactory delegate)
+    {
+        this(delegate, Duration.ofSeconds(2), 10);
+    }
+
+    ReusingConnectionFactory(ConnectionFactory delegate, Duration duration, long maximumSize)
+    {
+        requireNonNull(duration, "duration is null");
+        NonLoadableCache<String, Connection> connections = SafeCaches.buildNonLoadableCache(CacheBuilder.newBuilder()
+                .maximumSize(maximumSize)
+                .expireAfterWrite(duration)
+                .removalListener(notification -> {
+                    if (notification.getCause() == EXPLICIT) {
+                        // connection was taken from the cache
+                        return;
+                    }
+                    try {
+                        requireNonNull(notification.getValue(), "notification.getValue() is null");
+                        notification.getValue().close();
+                    }
+                    catch (SQLException e) {
+                        throw new TrinoException(JDBC_ERROR, e);
+                    }
+                }));
+        this.connections = connections.asMap();
+        this.delegate = requireNonNull(delegate, "delegate is null");
+    }
+
+    @Override
+    public Connection openConnection(ConnectorSession session)
+            throws SQLException
+    {
+        String queryId = session.getQueryId();
+        Connection connection = getConnection(session, queryId);
+        return new CachedConnection(queryId, connection);
+    }
+
+    private Connection getConnection(ConnectorSession session, String queryId)
+            throws SQLException
+    {
+        Connection connection = connections.remove(queryId);
```

Comment:
`.remove` I'd expect that connections holder (this class) is called `pool`

Reply:
it's not a typical pool
This is effectively a queryId-keyd cache of 1-element pools.
It should be documented by keying by queryId is used here. This is not obvious especially that we don't reuse any "dirty" connections.